initial attempt to implementing hard majority voting

LukasGardberg opened the pull request:
This is a first start to implementing hard majority voting, as mention in #118.

For a brief description of soft and hard voting, see [sklearn's definition](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html).

> voting{‘hard’, ‘soft’}, default=’hard’
> 
> If ‘hard’, uses predicted class labels for majority rule voting. Else if ‘soft’, predicts the class label based on the argmax of the sums of the predicted probabilities, which is recommended for an ensemble of well-calibrated classifiers.

For hard voting, the vector `proba` returned by the [`forward`](https://github.com/TorchEnsemble-Community/Ensemble-Pytorch/blob/master/torchensemble/voting.py#L99) method of the VotingClassifier class now instead becomes a one-hot vector with a 1 on the class which had the majority of the base_estimators votes. 

Worth considering is that `torch.max` is called in the [`evaluate`](https://github.com/TorchEnsemble-Community/Ensemble-Pytorch/blob/master/torchensemble/_base.py#L284) method which essentially just gets the index of the 1 in `proba` from the `forward` method, which works, but maybe isn't a problem?

Any suggestions for improvements are very welcome!

xuyxu commented:
Thanks @LukasGardberg, dont worry about the CI problem. I will leave the detailed comment as soon as possible.

xuyxu commented on torchensemble/_base.py:
```diff
@@ -280,8 +280,16 @@ def evaluate(self, test_loader, return_loss=False):
 
         for _, elem in enumerate(test_loader):
             data, target = split_data_target(elem, self.device)
+
```
There is no need to add these extra comments.

xuyxu commented on torchensemble/_base.py:
```diff
@@ -349,8 +357,8 @@ def __init__(self, input_dim, output_dim, depth=5, lamda=1e-3, cuda=False):
 
-        self.internal_node_num_ = 2 ** self.depth - 1
-        self.leaf_node_num_ = 2 ** self.depth
+        self.internal_node_num_ = 2**self.depth - 1
```
Several lines here do not meet the coding style requirements. Please consider to execute the command `black --skip-string-normalization --config pyproject.toml torchensemble/` in the root directory of the project, which will automatically format the code

LukasGardberg replied:
Weird, I ran the same `black` command and it automatically did the changes mentioned above. I reverted them manually now. Maybe it has something to do with my version of `black`.

xuyxu commented on torchensemble/voting.py:
```diff
@@ -92,17 +92,52 @@ def _parallel_fit_per_epoch(
+    def __init__(self, voting_strategy="soft", **kwargs):
+        super(VotingClassifier, self).__init__(**kwargs)
+
+        self._implemented_voting_strategies = ["soft", "hard"]
```
This attribute is redundant, maybe we could hard-code the supported strategies, for example, change the attribute in line 100 into a set `("soft", "hard")`, and replace the latter `{}` in line 103 with a fixed string.

LukasGardberg replied:
Check if my latest changes is what you meant :)

xuyxu commented on torchensemble/voting.py:
```diff
@@ -92,17 +92,52 @@ def _parallel_fit_per_epoch(
     def forward(self, *x):
         # Average over class distributions from all base estimators.
+
+        # output: (n_estimators, batch_size, n_classes)
```
No need to add this comment.

xuyxu reviewed (commented):
The core implementation on voting strategies looks good after a rough look, we need extra unit tests to ensure the correctrness. Have you ever written unit tests?

LukasGardberg commented:
Yes, a bit, but never with pytest. I'm thinking I'll just imitate the tests that already exists. Is there anything specific I should test for?

Also, when running the tests locally I seem to be getting an error:

` AttributeError: 'NeuralForestClassifier' object has no attribute 'voting_strategy'`

I tried adding an init method to the `NeuralForestClassifier` class, but am a bit unsure how to deal with having two superclasses, and how to properly handle this problem. Do you have any advice for how to fix it :)? Sorry for needing some guidance haha

xuyxu commented:
Thanks for your great work @LukasGardberg ! Kind of busy these days, and I will take a look in a few days.

xuyxu commented on torchensemble/voting.py:
```diff
@@ -92,16 +92,39 @@ def _parallel_fit_per_epoch(
+        if self.voting_strategy == "soft":
+            proba = op.average(outputs)
+
+        elif self.voting_strategy == "hard":
```
One thing comes to my mind is that aside from `voting`, parallel ensemble methods like `Snapshot Ensemble`, `Neural Forest` could all benefit from various voting strategies ;-).

Therefore, would it be convenient for you to move the functionality of hard voting to the `op` sub-package, and make all parallel ensemble methods support specifying the voting strategy manually?

xuyxu reviewed (commented):
As to unit tests, provided that we move the implementation of hard voting to the `utils/operator` sub-package, it should be enough to only test this function in `tests/test_operator.py`. You could manually specify several tensors, and see if the result from hard voting conforms to the expected result.

xuyxu commented:
Also, please click the `Details` button of the CI `torchensemble-CI / build (ubuntu-latest, 3.7) (pull_request)` and see pytest results, since the neural forest class inherits from voting, its `__init__` method should be modified accordingly.

xuyxu commented:
@all-contributors please add @LukasGardberg for code

LukasGardberg commented:
Okay, now hard majority has been moved to ops. I also added a simple test, and added soft voting to NeuralForestClassifier and SnapshotEnsembleClassifier, but hey have not been tested.

Are there any other models I should add hard voting for? I'm assuming it's only applicable to classifiers right?

Also, I'm unsure if I solved the init inheritance problem properly, but the tests are at least passing now @xuyxu :)

xuyxu commented on torchensemble/_base.py:
```diff
@@ -203,7 +203,7 @@ def predict(self, *x):
 class BaseTreeEnsemble(BaseModule):
     def __init__(
         self,
-        n_estimators,
```
Is there any reason on adding the default value for `n_estimators`

xuyxu commented on torchensemble/utils/operator.py:
```diff
@@ -51,3 +53,21 @@ def pseudo_residual_regression(target, output):
+    majority_one_hots: (n_samples, n_classes)
+    """
+
+    assert (
```
The `assert` clause is not recommended in the code for production, therefore, it would be better to use `raise ValueError` here ;-)

xuyxu reviewed (commented):
Please refer to the comments, and I will add some final docs once they are resolved.

xuyxu reviewed (commented):
LGTM.

xuyxu commented:
Merged, thanks @LukasGardberg

xuyxu commented:
@all-contributors please add @LukasGardberg for code

LukasGardberg commented:
Thanks a lot @xuyxu for all the help! Exciting making my first contribution :)
